addon/components/utils/base-popover.js

```diff
@@ -0,0 +1,219 @@
+      this._triggerElement.removeEventListener('mouseleave', this._closePopoverHandler);
+    }
+
+    this._triggerElement.removeAttribute('data-popover-trigger');
+
+    document.body.removeEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Inserts the popover and sets up the body click listener
+   */
+  _insertPopover() {
+    // Schedule these separately so that the element gets fully attached by setting
+    // `isOpen` to true, and only _then_ setting `isShowing` to true, triggering the
+    // CSS transition
+    run(() => this.set('isOpen', true));
+
+    document.body.addEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Removes the popover and tears down the body click listener
+   */
+  _removePopover() {
+    run(() => this.set('isOpen', false));
+
+    document.body.removeEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Triggers when the popover has entered the DOM and the API has been established,
+   * allowing us to add data attributes and event handlers and mark the trigger as active
+   *
+   * @param {PopperAPI} api - The API received from the underlying popper
+   */
+  @action
+  popoverOpened({ popperElement }) {
+    this._popperElement = popperElement;
+    this._popperElement.setAttribute('data-popover-content', guidFor(this));
+    this._triggerElement.classList.add('is-active');
+
+    if (this.get('triggerEvent') === 'hover') {
+      this._popperElement.addEventListener('mouseenter', this._openPopoverHandler);
+      this._popperElement.addEventListener('mouseleave', this._closePopoverHandler);
+    }
+  }
+
+  /**
+   * Triggers when the popover has been fully removed from the DOM (e.g. after animations)
+   * allowing us to teardown event handlers and mark the trigger element as inactive
+   */
+  @action
+  popoverClosed() {
+    if (this.get('triggerEvent') === 'hover') {
+      this._popperElement.removeEventListener('mouseenter', this._openPopoverHandler);
+      this._popperElement.removeEventListener('mouseleave', this._closePopoverHandler);
+    }
+
+    this._triggerElement.classList.remove('is-active');
+    this._popperElement.removeAttribute('data-popover-content');
+    this._popperElement = null;
+  }
+
+  // ----- Event Handlers -----
+
+  /**
+   * Opens the popover
+   */
+  _openPopoverHandler = () => {
+    if (!this.get('isOpen')) {
+      this._isOpening = true;
+      this._insertPopover();
+    }
+  }
+
+  /**
+   * Closes the popover
+   */
+  _closePopoverHandler = () => {
+    if (this.get('isOpen')) {
+      this._isOpening = false;
+      this._removePopover();
+    }
+  }
+
+  /**
+   * Handles a click on the body in general when a popover is open. If the clicked element is not
+   * a child of the popover, or is marked with `data-close`, closes the popover.
+   */
+  _handleBodyClick = ({ target }) => {
+    if (this._isOpening) {
+      this._isOpening = false;
+    } else if (closest(target, '.ice-base-popover') === null) {
```

Comment:
What happens if `target` is a deleted item inside popover? We are not supposed to close the popover in that case but the target is not inside the popover since Ember has removed it from DOM. 

We have seen this bug before. cc @LydiaBarnes

Reply:
That sounds like a timing issue within Ember, events should never be resolved _after_ changes due to a render. We should be able to mitigate this by fixing the architecture of the component in question.